site/bps/BP-38-bookie-endpoint-discovery.md

```diff
@@ -0,0 +1,149 @@
+        String[] extensions; // "starttls"
+    }
+    List<Endpoint> endpoints;
+    Map<String, String> properties; // additional properties
+}
+
+```
+
+In RegistrationClient interface we expose a new method:
+
+```
+CompletableFuture<Versioned<BookieServiceInfo>> getBookieServiceInfo(String bookieId)
+```
+
+The client can derive bookieId from a BookieSocketAddress. He can access the list of available bookies using **RegistrationClient#getAllBookies()** and then use this new method to get the details of the services exposed by the Bookie.
+
+On the Bookie side we change the RegistrationManager interface in order to let the Bookie
+advertise the services:
+
+in RegistrationManager class the **registerBookie** method 
+```
+void registerBookie(String bookieId, boolean readOnly)
+```
+
+becomes
+
+```
+void registerBookie(String bookieId, boolean readOnly, BookieServiceInfo bookieServieInfo)
+```
+
+It will be up to the implementation of RegistrationManager and RegistrationClient to serialize
+the BookieServiceInfo structure.
+
+For the ZooKeeper based implementation we are going to store such information in Protobuf binary format, as currently this is the format used for every other kind of metadata (the example here is in JSON-like for readability purposes):
+
+```
+{
+    "endpoints": [
+        {
+         "name": "bookie",
+         "hostname": "localhost",
+         "port": 3181,
+         "protocol": "bookie-rpc",
+         "auth": ["sasl-gsapi"],
+         "extensions": ["starttls"]
+        },
+        {
+         "name": "bookie-http-server",
+         "hostname": "localhost",
+         "port": 8080,
+         "protocol": "http",
+         "auth": [],
+         "extensions": []
+        }
+    ],
+    properties {
+        "property": "value",
+        "property2": "value2"
+    }
+}
+```
+Such information will be stored inside the '/REGISTRATIONPATH/available' znode (or /REGISTRATIONPATH/available/readonly' in case of readonly bookie), these paths are the same used in 4.10, but in 4.10 we are writing empty znodes.
+
+The rationale around this choice is that the client is already using these znodes in order to discover available bookies services.
+
+It is possible that such endpoint information change during the lifetime of a Bookie, like after rebooting a machine and the machine gets a new network address.
+
+It is out of the scope of this proposal to change semantics of ledger metadata, in which we  are currently storing directly the network address of the bookies, but with this proposal we are preparing the way to have an indirection and separate the concept of Bookie ID from the Network address.
+
+**Endpoint structure**
+- **name**: this is an id for the service type, like 'bookie'
+- **hostname**: the hostname (or a raw IP address)
+- **port**: the TCP port (int)
+- **protocol**: the type of service (bookie-rpc, http, https)
+- **auth**: supported authentication types (tls-auth, sasl-gssapi...)
```

Comment:
What is the difference between `auth` and `extensions`? Can you give an example of how they are used?

Reply:
@sijie 
**Auth** from my point of view describes how the bookie is able to perform *Authentication*, that is how to bind an user to the connection/request, it is useful in order to select Bookie for which the client has the request auth type available and for monitoring purposes (I want to see if all of the bookies are exposing the expected auth type)

**Extensions** is more generic and it will depends on the protocol, in the case of bookie-rpc we have the ability to use TLS, but it is only an _extension_ to the base protocol. In fact we are using the StartTLS mechanism, if we had a "pure" TLS connection I would add a new *protocol* named "bookie-rpcs" or something like that.

Reply:
regarding `Auth`, if it is only for monitoring purpose. it should be done by exposing metrics for that. If we are exposing this to the bookie service info, we should describe how clients are going to use this. If there is no clear use cases for using this, we should try to avoid adding them right now.

if you have a new protocol, that would be new endpoint, the endpoint name should already  be used for distinguishing different endpoints. I don't know we need an extension.

Reply:
Dropped *Auth* and *Extensions* from the document